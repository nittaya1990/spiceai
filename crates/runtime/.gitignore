abc-sqlite.db-wal
abc-sqlite.db-shm
