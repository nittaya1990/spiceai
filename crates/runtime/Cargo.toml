[package]
description = "The spiced runtime"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "runtime"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
app = { path = "../app" }
arrow.workspace = true
arrow-csv = "53.3.0"
arrow-flight = { workspace = true, features = ["flight-sql-experimental"] }
arrow-ipc.workspace = true
arrow-json.workspace = true
arrow-schema = { workspace = true, features = ["serde"] }
arrow_sql_gen = { path = "../arrow_sql_gen" }
arrow_tools = { path = "../arrow_tools" }
async-openai.workspace = true
async-stream.workspace = true
async-trait.workspace = true
aws-config = { version = "1.1.10", optional = true }
aws-sdk-dynamodb = { version = "1", optional = true }
aws-sdk-secretsmanager = { version = "1", optional = true }
aws-sdk-sts = { version = "1", optional = true }
axum.workspace = true
axum-extra = { version = "0.9.6", features = ["typed-header"] }
base64.workspace = true
bytes.workspace = true
cache = { path = "../cache" }
chrono = { version = "0.4.38" }
clap.workspace = true
clickhouse-rs = { workspace = true, optional = true }
csv = "1.3.0"
dashmap = "6.1.0"
data_components = { path = "../data_components" }
datafusion.workspace = true
datafusion-federation = { workspace = true }
datafusion-federation-sql = { workspace = true }
datafusion-functions-json = { workspace = true }
datafusion-table-providers = { workspace = true }
db_connection_pool = { path = "../db_connection_pool" }
document_parse = { path = "../document_parse" }
dotenvy.workspace = true
duckdb = { workspace = true, features = [
    "bundled",
    "r2d2",
    "vtab",
    "vtab-arrow",
], optional = true }
flight_client = { path = "../flight_client" }
fundu = { workspace = true }
futures.workspace = true
globset.workspace = true
governor = "0.8.0"
graph-rs-sdk = { workspace = true, optional = true }
graphql-parser.workspace = true
headers-accept = "0.1.3"
http.workspace = true
http-body-util = "0.1.2"
hyper = "1.6.0"
hyper-util = { version = "0.1.6", features = ["service"] }
iceberg.workspace = true
iceberg-catalog-rest.workspace = true
imap = { workspace = true, optional = true }
indexmap.workspace = true
itertools.workspace = true
jsonwebtoken = "9.3.0"
keyring = { version = "3.6.1", features = [
    "apple-native",
    "windows-native",
    "linux-native",
], optional = true }
llms = { path = "../llms" }
logos = "0.15.0"
mcp-client = { git = "https://github.com/block/goose", package = "mcp-client", optional = true }
mcp-core = { git = "https://github.com/block/goose", package = "mcp-core", optional = true }
mediatype = "0.19.18"
model_components = { path = "../model_components" }
mysql_async = { workspace = true, optional = true }
notify = "7.0.0"
ns_lookup = { path = "../ns_lookup" }
object_store = { workspace = true, features = ["aws", "http", "azure"] }
once_cell = "1.19.0"
opentelemetry.workspace = true
opentelemetry-prometheus.workspace = true
opentelemetry-proto = { version = "0.27", features = [
    "gen-tonic-messages",
    "gen-tonic",
    "metrics",
] }
opentelemetry_sdk.workspace = true
otel-arrow = { path = "../otel-arrow" }
package = { path = "../package" }
percent-encoding.workspace = true
pin-project.workspace = true
prometheus.workspace = true
prometheus-parse = "0.2.5"
prost = { version = "0.13.1", default-features = false, features = [
    "prost-derive",
] }
rand = "0.8.5"
regex.workspace = true
reqwest.workspace = true
runtime-auth = { path = "../runtime-auth" }
rusqlite = { workspace = true, optional = true }
rustls.workspace = true
rustls-pemfile.workspace = true
schemars = "0.8.19"
secrecy.workspace = true
serde = { workspace = true, features = ["derive"] }
serde_json.workspace = true
snafu.workspace = true
snowflake-api = { workspace = true, optional = true }
spicepod = { path = "../spicepod" }
ssh2 = { workspace = true, optional = true }
suppaftp = { workspace = true, optional = true }
telemetry = { path = "../telemetry" }
tiberius = { workspace = true, optional = true }
tokio.workspace = true
tokio-rusqlite = { workspace = true, optional = true }
tokio-rustls = "0.26.0"
tokio-stream.workspace = true
tokio-util = { workspace = true, optional = true }
tonic.workspace = true
tonic-health.workspace = true
tower.workspace = true
tower-http.workspace = true
tracing.workspace = true
tracing-futures.workspace = true
tract-core = "0.21.0"
url = "2.5.4"
util = { path = "../util" }
utoipa = { version = "5", features = ["axum_extras"], optional = true }
utoipa-swagger-ui = { version = "8", features = ["axum"], optional = true }

uuid = { workspace = true, features = ["v7"] }
x509-certificate.workspace = true

[dev-dependencies]
anyhow = "1.0.86"
async-graphql = "7.0.5"
async-graphql-axum = "=7.0.13"
azure_core = "0.21.0"
azure_storage = "0.21.0"
azure_storage_blobs = "0.21.0"
bollard = "0.18.1"
dirs = "5.0.1"
flightrepl = { path = "../flightrepl" }
hf-hub = { workspace = true }
insta.workspace = true
jsonpath-rust = "0.7.3"
opentelemetry_sdk = { workspace = true, features = ["rt-tokio-current-thread"] }
paste.workspace = true
scopeguard = "1.2.0"
serde_json = { workspace = true, features = ["preserve_order"] }
spice-cloud = { path = "../spice_cloud" }
test-framework = { path = "../test-framework" }
tokio = { workspace = true, features = ["time", "test-util"] }
tracing-opentelemetry.workspace = true
tracing-subscriber.workspace = true

[features]
aws-secrets-manager = [
    "dep:aws-config",
    "dep:aws-sdk-secretsmanager",
    "dep:aws-sdk-sts",
]
clickhouse = [
    "dep:clickhouse-rs",
    "db_connection_pool/clickhouse",
    "arrow_sql_gen/clickhouse",
    "data_components/clickhouse",
]
cuda = ["llms/cuda"]
databricks = ["data_components/databricks"]
debezium = ["data_components/debezium"]
default = ["keyring-secret-store", "aws-secrets-manager", "sharepoint"]
delta_lake = ["data_components/delta_lake"]
dev = ["openapi", "otel-arrow/dev", "mcp"]
dremio = []
duckdb = ["dep:duckdb", "db_connection_pool/duckdb", "data_components/duckdb"]
dynamodb = ["dep:aws-sdk-dynamodb", "data_components/dynamodb"]
flightsql = ["data_components/flightsql"]
ftp = ["dep:suppaftp", "dep:ssh2"]
imap = ["data_components/imap", "dep:imap"]
keyring-secret-store = ["dep:keyring"]
mcp = ["dep:mcp-client", "dep:mcp-core"]
metal = ["llms/metal"]
models = ["model_components/full"]
mssql = ["dep:tiberius", "dep:tokio-util", "data_components/mssql"]
mysql = ["dep:mysql_async", "db_connection_pool/mysql", "data_components/mysql"]
odbc = ["db_connection_pool/odbc", "data_components/odbc"]
openapi = ["dep:utoipa-swagger-ui", "dep:utoipa"]
postgres = [
    "db_connection_pool/postgres",
    "data_components/postgres",
    "datafusion-table-providers/postgres",
]
postgres-write = []
sharepoint = ["data_components/sharepoint", "dep:graph-rs-sdk"]
snowflake = [
    "dep:snowflake-api",
    "db_connection_pool/snowflake",
    "data_components/snowflake",
]
spark = ["data_components/spark_connect"]
sqlite = [
    "dep:tokio-rusqlite",
    "db_connection_pool/sqlite",
    "data_components/sqlite",
    "dep:rusqlite",
]

[[bench]]
harness = false
name = "bench"

[[bench]]
harness = false
name = "vector_search"

[target.'cfg(windows)'.dependencies]
winver = "1.0.0"
