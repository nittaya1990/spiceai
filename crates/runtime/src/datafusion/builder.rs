/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{
    collections::HashSet,
    sync::{Arc, RwLock},
};

use cache::QueryResultsCacheProvider;
use datafusion::{
    catalog_common::{CatalogProvider, MemoryCatalogProvider},
    execution::SessionStateBuilder,
    optimizer::{
        analyzer::{
            count_wildcard_rule::CountWildcardRule, expand_wildcard_rule::ExpandWildcardRule,
            inline_table_scan::InlineTableScan, resolve_grouping_function::ResolveGroupingFunction,
            type_coercion::TypeCoercion,
        },
        AnalyzerRule,
    },
    prelude::{SessionConfig, SessionContext},
};
use datafusion_federation::FederationAnalyzerRule;
use tokio::sync::RwLock as TokioRwLock;

use crate::{embeddings, object_store_registry::default_runtime_env, status};

use super::{
    extension::{bytes_processed::BytesProcessedOptimizerRule, SpiceQueryPlanner},
    schema::SpiceSchemaProvider,
    DataFusion, SPICE_DEFAULT_CATALOG, SPICE_DEFAULT_SCHEMA, SPICE_METADATA_SCHEMA,
    SPICE_RUNTIME_SCHEMA,
};

pub struct DataFusionBuilder {
    config: SessionConfig,
    status: Arc<status::RuntimeStatus>,
    cache_provider: Option<Arc<QueryResultsCacheProvider>>,
}

pub(crate) fn get_df_default_config() -> SessionConfig {
    let mut df_config = SessionConfig::new();

    // Prevents DataFusion from lowercasing identifiers, i.e. "SELECT MyColumn FROM my_table" would be "SELECT mycolumn FROM mytable" without this.
    // This improves the UX for data sources where column names are case-sensitive, since they no longer need to be quoted.
    df_config
        .options_mut()
        .sql_parser
        .enable_ident_normalization = false;

    df_config.options_mut().optimizer.expand_views_at_output = true;
    df_config.options_mut().sql_parser.dialect = "PostgreSQL".to_string();
    df_config
        .options_mut()
        .execution
        .listing_table_ignore_subdirectory = false;

    // There are some unidentified bugs in DataFusion that cause schema checks to fail for aggregate functions.
    // Spice is affected by this - skip the check until all bugs are fixed.
    // Tracking issue: https://github.com/apache/datafusion/issues/12733
    df_config
        .options_mut()
        .execution
        .skip_physical_aggregate_schema_check = true;

    df_config
}

impl DataFusionBuilder {
    #[must_use]
    pub fn new(status: Arc<status::RuntimeStatus>) -> Self {
        let mut df_config = get_df_default_config()
            .with_information_schema(true)
            .with_create_default_catalog_and_schema(false);

        df_config.options_mut().catalog.default_catalog = SPICE_DEFAULT_CATALOG.to_string();
        df_config.options_mut().catalog.default_schema = SPICE_DEFAULT_SCHEMA.to_string();

        Self {
            config: df_config,
            status,
            cache_provider: None,
        }
    }

    #[must_use]
    pub fn with_cache_provider(mut self, cache_provider: Arc<QueryResultsCacheProvider>) -> Self {
        self.cache_provider = Some(cache_provider);
        self
    }

    /// Builds the `DataFusion` instance.
    ///
    /// # Panics
    ///
    /// Panics if the `DataFusion` instance cannot be built due to errors in registering functions or schemas.
    #[must_use]
    pub fn build(self) -> DataFusion {
        let mut state = SessionStateBuilder::new()
            .with_config(self.config)
            .with_default_features()
            .with_query_planner(Arc::new(SpiceQueryPlanner::new()))
            .with_runtime_env(default_runtime_env())
            .with_analyzer_rules(get_analyzer_rules())
            .build();

        if let Err(e) = datafusion_functions_json::register_all(&mut state) {
            panic!("Unable to register JSON functions: {e}");
        };

        let ctx = SessionContext::new_with_state(state);
        ctx.add_optimizer_rule(Arc::new(BytesProcessedOptimizerRule::new()));
        ctx.register_udf(embeddings::cosine_distance::CosineDistance::new().into());
        ctx.register_udf(crate::datafusion::udf::Greatest::new().into());
        ctx.register_udf(crate::datafusion::udf::Least::new().into());
        let catalog = MemoryCatalogProvider::new();
        let default_schema = SpiceSchemaProvider::new();
        let runtime_schema = SpiceSchemaProvider::new();

        let metadata_schema = SpiceSchemaProvider::new();

        match catalog.register_schema(SPICE_DEFAULT_SCHEMA, Arc::new(default_schema)) {
            Ok(_) => {}
            Err(e) => {
                panic!("Unable to register default schema: {e}");
            }
        }

        match catalog.register_schema(SPICE_RUNTIME_SCHEMA, Arc::new(runtime_schema)) {
            Ok(_) => {}
            Err(e) => {
                panic!("Unable to register spice runtime schema: {e}");
            }
        }

        if cfg!(feature = "models") {
            use super::SPICE_EVAL_SCHEMA;
            let eval_schema = SpiceSchemaProvider::new();
            match catalog.register_schema(SPICE_EVAL_SCHEMA, Arc::new(eval_schema)) {
                Ok(_) => {}
                Err(e) => {
                    panic!("Unable to register spice eval schema: {e}");
                }
            }
        }

        match catalog.register_schema(SPICE_METADATA_SCHEMA, Arc::new(metadata_schema)) {
            Ok(_) => {}
            Err(e) => {
                panic!("Unable to register spice runtime schema: {e}");
            }
        }

        ctx.register_catalog(SPICE_DEFAULT_CATALOG, Arc::new(catalog));

        DataFusion {
            runtime_status: self.status,
            ctx: Arc::new(ctx),
            data_writers: RwLock::new(HashSet::new()),
            cache_provider: RwLock::new(self.cache_provider),
            pending_sink_tables: TokioRwLock::new(Vec::new()),
            accelerated_tables: TokioRwLock::new(HashSet::new()),
        }
    }
}

/// Spice customizes the order of the analyzer rules, since some of them are only relevant when `DataFusion` is executing the query,
/// as opposed to when underlying federated query engines will execute the query.
///
/// This list should be kept in sync with the default rules in `Analyzer::new()`, but with the federation analyzer rule added.
fn get_analyzer_rules() -> Vec<Arc<dyn AnalyzerRule + Send + Sync>> {
    vec![
        Arc::new(InlineTableScan::new()),
        Arc::new(ExpandWildcardRule::new()),
        Arc::new(FederationAnalyzerRule::new()),
        // The rest of these rules are run after the federation analyzer since they only affect internal DataFusion execution.
        Arc::new(ResolveGroupingFunction::new()),
        Arc::new(TypeCoercion::new()),
        Arc::new(CountWildcardRule::new()),
    ]
}

#[cfg(test)]
mod tests {
    use datafusion::optimizer::Analyzer;

    /// Verifies that the default analyzer rules are in the expected order.
    ///
    /// If this test fails, `DataFusion` has modified the default analyzer rules and `get_analyzer_rules()` should be updated.
    #[test]
    fn test_verify_default_analyzer_rules() {
        let default_rules = Analyzer::new().rules;
        assert_eq!(
            default_rules.len(),
            5,
            "Default analyzer rules have changed"
        );
        let expected_rule_names = vec![
            "inline_table_scan",
            "expand_wildcard_rule",
            "resolve_grouping_function",
            "type_coercion",
            "count_wildcard_rule",
        ];
        for (rule, expected_name) in default_rules.iter().zip(expected_rule_names.into_iter()) {
            assert_eq!(
                expected_name,
                rule.name(),
                "Default analyzer rule order has changed"
            );
        }
    }
}
