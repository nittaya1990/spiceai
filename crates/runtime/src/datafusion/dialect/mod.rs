/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::sync::Arc;

use datafusion::sql::unparser::dialect::{Dialect, DuckDBDialect, ScalarFnToSqlHandler};

mod duckdb;

/// Creates a new instance of the `DuckDB` dialect with support for Spice internal UDFs
pub fn new_duckdb_dialect() -> Arc<dyn Dialect> {
    let dialect = DuckDBDialect::new().with_custom_scalar_overrides(vec![(
        "cosine_distance",
        Box::new(duckdb::cosine_distance_to_sql) as ScalarFnToSqlHandler,
    )]);

    Arc::new(dialect) as Arc<dyn Dialect>
}
