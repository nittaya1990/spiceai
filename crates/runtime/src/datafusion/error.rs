/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//! Spice specific errors that are returned as part of `DataFusionError::External`.

use std::{fmt::Display, sync::Arc};

use arrow_schema::ArrowError;
use datafusion::error::DataFusionError;
use datafusion_table_providers::util::retriable_error::RetriableError;

#[derive(Debug)]
pub enum SpiceExternalError {
    AccelerationNotReady { dataset_name: String },
}

impl SpiceExternalError {
    #[must_use]
    pub fn acceleration_not_ready(
        dataset_name: String,
    ) -> Box<dyn std::error::Error + Send + Sync> {
        Box::new(Self::AccelerationNotReady { dataset_name })
    }
}

impl std::error::Error for SpiceExternalError {}

impl Display for SpiceExternalError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SpiceExternalError::AccelerationNotReady { dataset_name } => write!(
                f,
                "Acceleration not ready; loading initial data for {dataset_name}"
            ),
        }
    }
}

#[must_use]
pub fn get_spice_df_error(e: &DataFusionError) -> Option<&SpiceExternalError> {
    match e {
        DataFusionError::External(e) => {
            if let Some(spice_err) = e.downcast_ref::<SpiceExternalError>() {
                Some(spice_err)
            } else if let Some(retry_err) = e.downcast_ref::<RetriableError>() {
                match retry_err {
                    RetriableError::DataRetrievalError {
                        source: DataFusionError::External(e),
                    } => e.downcast_ref::<SpiceExternalError>(),
                    _ => None,
                }
            } else {
                None
            }
        }
        _ => None,
    }
}

/// Finds the root `DataFusionError` if multiple errors are nested, maintaining ownership of the `DataFusionError`.
/// We implement a custom matcher because we need to retain ownership of the `DataFusionError`, but `DataFusionError::find_root()` returns a reference.
///
/// This function does not unnest `Box<Arc<DataFusionError>>` because it cannot take ownership of the inner `DataFusionError`.
#[must_use]
pub fn find_datafusion_root(e: DataFusionError) -> DataFusionError {
    let mut last_error = e;

    tracing::debug!("Finding root of DataFusionError: {:?}", last_error);

    loop {
        match last_error {
            DataFusionError::External(err) => match err.downcast::<DataFusionError>() {
                Ok(inner) => last_error = *inner,
                Err(err) => match err.downcast::<Arc<DataFusionError>>() {
                    Ok(inner) => {
                        tracing::debug!("Found Arc<DataFusionError> in External: {:?}", inner);
                        return DataFusionError::External(inner);
                    }
                    Err(err) => return DataFusionError::External(err),
                },
            },
            DataFusionError::Context(_, err) => last_error = *err,
            DataFusionError::ArrowError(ArrowError::ExternalError(err), message) => {
                match err.downcast::<DataFusionError>() {
                    Ok(inner) => last_error = *inner,
                    Err(err) => match err.downcast::<Arc<DataFusionError>>() {
                        Ok(inner) => {
                            tracing::debug!(
                                "Found Arc<DataFusionError> in ArrowError: {:?}",
                                inner
                            );
                            return DataFusionError::ArrowError(
                                ArrowError::ExternalError(inner),
                                message,
                            );
                        }
                        Err(err) => {
                            return DataFusionError::ArrowError(
                                ArrowError::ExternalError(err),
                                message,
                            )
                        }
                    },
                }
            }
            _ => return last_error,
        }
    }
}
