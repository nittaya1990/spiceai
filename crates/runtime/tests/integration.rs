/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#![allow(clippy::large_futures)]

use std::sync::Arc;

use arrow::{array::RecordBatch, util::display::FormatOptions};
use datafusion::{
    execution::context::SessionContext,
    parquet::arrow::arrow_reader::ParquetRecordBatchReaderBuilder,
};
use futures::TryStreamExt;

use runtime::{datafusion::DataFusion, status, Runtime};
use tracing::subscriber::DefaultGuard;
use tracing_subscriber::EnvFilter;
mod abfs;
mod acceleration;
mod catalog;
mod cors;
#[cfg(all(feature = "delta_lake", feature = "databricks"))]
mod databricks_delta;
#[cfg(all(feature = "delta_lake", feature = "databricks"))]
mod databricks_delta_catalog;
#[cfg(all(feature = "spark", feature = "databricks"))]
mod databricks_spark;
#[cfg(all(feature = "spark", feature = "databricks"))]
mod databricks_spark_catalog;
#[cfg(feature = "delta_lake")]
mod delta_lake;
mod docker;
#[cfg(feature = "duckdb")]
mod duckdb;
mod endpoint_auth;
mod file;
mod flight;
mod github;
mod graphql;
#[cfg(feature = "mssql")]
mod mssql;
#[cfg(feature = "mysql")]
mod mysql;
#[cfg(feature = "odbc")]
mod odbc;
#[cfg(feature = "postgres")]
mod postgres;
mod refresh_retry;
mod refresh_sql;
mod results_cache;
mod s3;
#[cfg(feature = "snowflake")]
mod snowflake;
#[cfg(feature = "spark")]
mod spark;
mod spiceai;
#[cfg(feature = "sqlite")]
mod sqlite;
mod tls;
mod utils;

// MySQL is required for the rehydration tests
#[cfg(feature = "mysql")]
mod rehydration;

/// Gets a test `DataFusion` to make test results reproducible across all machines.
///
/// 1) Sets the number of `target_partitions` to 3, by default its the number of CPU cores available.
fn get_test_datafusion(status: Arc<status::RuntimeStatus>) -> Arc<DataFusion> {
    let mut df = DataFusion::builder(status).build();

    // Set the target partitions to 3 to make RepartitionExec show consistent partitioning across machines with different CPU counts.
    let mut new_state = df.ctx.state();
    new_state
        .config_mut()
        .options_mut()
        .execution
        .target_partitions = 3;
    let new_ctx = SessionContext::new_with_state(new_state);

    // Replace the old context with the modified one
    df.ctx = new_ctx.into();
    Arc::new(df)
}

fn init_tracing(default_level: Option<&str>) -> DefaultGuard {
    let filter = match (default_level, std::env::var("SPICED_LOG").ok()) {
        (_, Some(log)) => EnvFilter::new(log),
        (Some(level), None) => EnvFilter::new(level),
        _ => EnvFilter::new(
            "runtime=TRACE,datafusion-federation=TRACE,datafusion-federation-sql=TRACE",
        ),
    };

    let subscriber = tracing_subscriber::FmtSubscriber::builder()
        .with_env_filter(filter)
        .with_ansi(true)
        .finish();
    tracing::subscriber::set_default(subscriber)
}

async fn get_tpch_lineitem() -> Result<Vec<RecordBatch>, anyhow::Error> {
    let lineitem_parquet_bytes =
        reqwest::get("https://public-data.spiceai.org/tpch_lineitem.parquet")
            .await?
            .bytes()
            .await?;

    let parquet_reader =
        ParquetRecordBatchReaderBuilder::try_new(lineitem_parquet_bytes)?.build()?;

    Ok(parquet_reader.collect::<Result<Vec<_>, arrow::error::ArrowError>>()?)
}

type ValidateFn = dyn FnOnce(Vec<RecordBatch>);

async fn run_query_and_check_results<F>(
    rt: &mut Runtime,
    snapshot_name: &str,
    query: &str,
    snapshot_plan: bool,
    validate_result: Option<F>,
) -> Result<(), String>
where
    F: FnOnce(Vec<RecordBatch>),
{
    // Check the plan
    let query_results = rt
        .datafusion()
        .query_builder(&format!("EXPLAIN {query}"))
        .build()
        .run()
        .await
        .map_err(|e| format!("query `{query}` to plan: {e}"))?;

    let plan_results: Vec<RecordBatch> = query_results
        .data
        .try_collect::<Vec<RecordBatch>>()
        .await
        .map_err(|e| format!("query `{query}` to results: {e}"))?;

    println!("Query: {query}");

    let Ok(explain_plan) = arrow::util::pretty::pretty_format_batches(&plan_results) else {
        panic!("Failed to format plan");
    };

    if snapshot_plan {
        insta::with_settings!({
            description => format!("Query: {query}"),
            omit_expression => true
        }, {
            insta::assert_snapshot!(snapshot_name, explain_plan);
        });
    }

    // Check the result
    if let Some(validate_result) = validate_result {
        let result_batches = rt
            .datafusion()
            .query_builder(query)
            .build()
            .run()
            .await
            .map_err(|e| format!("query `{query}` failed to run: {e}"))?
            .data
            .try_collect()
            .await
            .map_err(|e| format!("query `{query}` to results: {e}"))?;

        validate_result(result_batches);
    }

    Ok(())
}

type PlanCheckFn = Box<dyn Fn(&str) -> bool>;

async fn run_query_and_check_results_with_plan_checks<F>(
    rt: &mut Runtime,
    query: &str,
    expected_plan_checks: Vec<(&str, PlanCheckFn)>,
    validate_result: Option<F>,
) -> Result<(), String>
where
    F: FnOnce(Vec<RecordBatch>),
{
    // Check the plan
    let plan_results: Vec<RecordBatch> = rt
        .datafusion()
        .query_builder(&format!("EXPLAIN {query}"))
        .build()
        .run()
        .await
        .map_err(|e| format!("query `{query}` to plan: {e}"))?
        .data
        .try_collect()
        .await
        .map_err(|e| format!("query `{query}` to results: {e}"))?;

    let Ok(formatted) = arrow::util::pretty::pretty_format_batches_with_options(
        &plan_results,
        &FormatOptions::default(),
    ) else {
        panic!("Failed to format plan");
    };
    let formatted = formatted.to_string();

    let actual_lines: Vec<&str> = formatted.trim().lines().collect();

    let mut matched_checks = vec![false; expected_plan_checks.len()];

    for line in actual_lines {
        for (i, (key, check_fn)) in expected_plan_checks.iter().enumerate() {
            if line.contains(key) {
                if matched_checks[i] {
                    return Err(format!(
                        "Check '{key}' matched multiple lines in plan:\n{formatted}",
                    ));
                }
                matched_checks[i] = true;
                if !check_fn(line) {
                    return Err(format!("Check failed for line: {line}"));
                }
            }
        }
    }

    if let Some(i) = matched_checks.iter().position(|&x| !x) {
        return Err(format!(
            "Expected check '{}' did not appear in plan:\n{formatted}",
            expected_plan_checks[i].0,
        ));
    }

    // Check the result
    if let Some(validate_result) = validate_result {
        let result_batches = rt
            .datafusion()
            .query_builder(query)
            .build()
            .run()
            .await
            .map_err(|e| format!("query `{query}` failed to run: {e}"))?
            .data
            .try_collect()
            .await
            .map_err(|e| format!("query `{query}` to results: {e}"))?;

        validate_result(result_batches);
    }

    Ok(())
}

fn container_registry() -> String {
    std::env::var("CONTAINER_REGISTRY")
        .unwrap_or_else(|_| "public.ecr.aws/docker/library/".to_string())
}
