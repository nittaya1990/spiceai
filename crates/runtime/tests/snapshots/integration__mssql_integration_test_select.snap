---
source: crates/runtime/tests/integration.rs
description: "Query: SELECT * FROM test"
snapshot_kind: text
---
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                             |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | BytesProcessedNode                                                                                                                                                                                                                               |
|               |   TableScan: test projection=[id, col_bit, col_tiny, col_short, col_long, col_longlong, col_float, col_double, col_timestamp, col_datetime, col_date, col_time, col_blob, col_varchar, col_string, col_var_string, col_decimal, col_char]        |
| physical_plan | BytesProcessedExec                                                                                                                                                                                                                               |
|               |   SqlServerExec sql=SELECT id, col_bit, col_tiny, col_short, col_long, col_longlong, col_float, col_double, col_timestamp, col_datetime, col_date, col_time, col_blob, col_varchar, col_string, col_var_string, col_decimal, col_char FROM test  |
|               |                                                                                                                                                                                                                                                  |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
